lib/routes/routeBackbeat.js

```diff
@@ -754,22 +794,94 @@ function deleteObjectTagging(request, response, log, callback) {
+        },
+    });
+}
+
+function _azureConditionalDelete(request, response, log, cb) {
+    const objectGetInfo = _createAzureConditionalDeleteObjectGetInfo(request);
+    const reqUids = log.getSerializedUids();
+    return dataClient.delete(objectGetInfo, reqUids, err => {
+        if (err) {
+            log.error('error deleting object in Azure', {
+                error: err,
+                method: '_azureConditionalDelete',
+            });
+            return cb(err);
+        }
+        return _respond(response, {}, log, cb);
+    });
+}
+
+function _conditionalTagging(request, response, locations, log, cb) {
+    return _getLastModified(locations, log, (err, lastModified) => {
+        if (err) {
+            return cb(err);
+        }
+        const ifUnmodifiedSince = request.headers['if-unmodified-since'];
+        if (ifUnmodifiedSince !== lastModified) {
+            log.info('object has been modified, skipping tagging operation', {
+                method: '_conditionalTagging',
+                ifUnmodifiedSince,
+                lastModified,
+            });
+            return cb();
+        }
+        return handleTaggingOperation(
+            request, response, 'Put', undefined, log, err => {
+                if (err) {
+                    log.error('put tagging failed', {
+                        method: '_conditionalTagging',
+                        error: err,
+                    });
+                    return cb(err);
+                }
+                return cb();
+            });
+    });
+}
+
+function _conditionalDelete(request, response, locations, log, cb) {
+    const { headers } = request;
+    const location = locationConstraints[headers['x-scal-storage-class']];
+    // Azure supports a conditional delete operation.
+    if (location && location.type === 'azure') {
+        return _azureConditionalDelete(request, response, log, cb);
+    }
+    // Other clouds do not support a conditional delete. Instead, we
+    // conditionally put tags to indicate if it should be deleted by the user.
+    return _conditionalTagging(request, response, locations, log, cb);
+}
+
 function batchDelete(request, response, log, callback) {
     return _getRequestPayload(request, (err, payload) => {
         if (err) {
             return callback(err);
         }
-        let request;
+        let parsedPayload;
         try {
-            request = JSON.parse(payload);
+            parsedPayload = JSON.parse(payload);
         } catch (e) {
             // FIXME: add error type MalformedJSON
             return callback(errors.MalformedPOSTRequest);
         }
-        if (!request || !Array.isArray(request.Locations)) {
+        if (!parsedPayload || !Array.isArray(parsedPayload.Locations)) {
             return callback(errors.MalformedPOSTRequest);
         }
-        const locations = request.Locations;
+        const locations = parsedPayload.Locations;
+        if (request.headers['if-unmodified-since']) {
```

Comment:
It seems if we ask to delete a specific version but also provide the "if-unmodified-since" header, it will delete the master version instead. I think it should not happen for transition policies but is potentially dangerous, we should always pass the version to delete if it was specified in the location IMHO.

Reply:
Ignore this, it should be fine here as we pass the locations array, but I think I'll have another related comment regarding versioning and put tagging :)

Reply:
It's conceived that the version check happens in backbeat. Sound like what you're suggesting is an explicit safety check for a version id in the `locations`? That makes sense to me, but just want to confirm.